### PR TITLE
Gitify

### PR DESCRIPTION
**Pull request by mariux**
beesh: Enable versionless bee filenames by setting #@beepackage in beefile

use '#@beepackage pkg-1.0-0' to set version information within bee file.

the bee file needs to match <PKGNAME>.bee or <PKGFULLNAME>.bee

This enables bee-files to be stored in git repositories and track changes.

### CODE REVIEW

**Comment by buczek** [comment]
This would be an alternate approach to what we did in https://github.molgen.mpg.de/mariux64/bee/commit/0d5deda08545dbb3809bcea3d21b120cea448168 which enhances the ``beeversion`` program only. At first we tried it in a way more similar to this one ( https://github.molgen.mpg.de/mariux64/bee/commit/8f399621fee1c303f11772380cde1414b912a947 ) but reverted, as I was not happy with the feature.

For example, we want a certain repository branch of bee files to match the files installed on a certain system. So a "pull request" translates to a "suggestion to install". A modification of the branch can be automated to an installation/deinstallation/upgrade. And of course we want to be able to verify, whether the files in the branch match the packages installed.

So you need an easy way to get the versions from the files checked out from the branch. You certainly don't want to run beesh on them, so you need to have the code which greps the version from the filename or the file content somewhere else anyway. The logical place is ``beeversion``. As it turns out, once you did that, you no longer need any changes on beesh or any other code.

We decided to make it depended on a the special filename suffix (``PKG.be0``)  . Logically we see .be0 files more as a source to be preprocessed into a bee-Files (``PKG-version.bee``) .This is much cleaner, as 'be0' is another language, in which the comment has a meaning, then 'bee' in which is has not. The package name and version of a 'be0' files is defined in its content, while in a 'bee' file it is defined in its name. The 'bee' format is used in BEE_BEEDIR or BEE_METADIR. The source file to be executed by beesh may be 'bee' or 'be0' format.

File: src/beesh.sh.in [hunk 1]
```diff
@@ -141,22 +141,20 @@ function check_rebuild() {
...
 
     beefile_pattern="${PKGFULLNAME}-${PKGFULLVERSION}"
     beefile_pattern=${beefile_pattern//+/\\+} # replace all + with \+
     beefile_pattern=${beefile_pattern//./\\.} # replace all . with \.
 
-    beefile_max=$(for bf in $(ls "${beefile%/*}/${beefile_pattern}*.bee" "${BEE_BEEDIR}/${beefile_pattern}*.bee" 2>/dev/null |\
+    beefile_max=$(for bf in $(ls ${beefile%/*}/${beefile_pattern}*.bee ${BEE_BEEDIR}/${beefile_pattern}*.bee 2>/dev/null |\
                               grep -E "/${beefile_pattern}-[0-9]+.bee") ; do
                       echo "${bf}"
                   done | ${BEESORT} | tail -1)
```
Comment (buczek) [inline src/beesh.sh.in]: consider https://github.molgen.mpg.de/mariux64/bee/commit/118648c2d80e2c612113076d66e60d4602b31cd7

**Comment by mariux** [comment]
the current .be0 implementation has a lot of implementation issues.. 

but it's ok to add it to beeversion for .be0 files... if this is what you think you need.. (i do not see another language here [same interpreter, same content, just one comment added] ... but implemented in beeversion you need this to not read all files - that is the actual reason for the suffix change... isn't it? - i don't think the new suffix is needed - but it does not break anything.. so: why not? ;) ...but i would suggest a small create/convert beefiles script with real versions... because the usecase is very special and needs additional tools anyway..)

but beeversion should check if package basename (full and normal) matches be0 filename.. and the comment should be much clearer (see below).. also it should go into beeversion_parse.c and not be hacked into main. In addition ends_with() lacks return value for failure.

i would suggest something like "#@beepackage"  instead of "#@beeversion" (this pull request) and/or  "# BEE_VERSION" (.be0 implementation) to make clear it needs to be a full beepackagename including all version information. (also more #@ pragma-like hints could be added in the future to manipulate runtime)

your implementation would still need changes from this commit to work without regressions.. (e.g. check existing beefiles)

also only the first appearance "^#@beepackage"  should "win" ..

File: src/beesh.sh.in (hunk 1 above)
Comment (mariux) [inline src/beesh.sh.in]: applied... thx

**Comment by buczek** [comment]
> you need this to not read all files - that is the actual reason for the suffix change... isn't it?

Not really. More to have a defined meaning if filename and comment don't match.

> but beeversion should check if package basename (full and normal) matches be0 filename

Not wanted by us. Junk in, junk out is fine. But we might want to do that on purpose some day and don't want a "you cant do that" reply. Also, things get complicated with packages like ``nvidia_linux-4.6.4-93-367.18`` which beeversion parses to ``PKGFULLNAME=nvidia_linux-4.6.4-93`` When version/revision is stripped (to ``nvidia_linux-4.6.4-93``) it is parsed to `` PKGFULLNAME=nvidia_linux``. With ``nvidia_linux-4.6.4-93-367``, which makes some sense, its parsed to ``PKGFULLNAME=nvidia_linux-4.6.4`` which is also not what we want.  Without the check, we can ignore all this trouble. We can name the file as we see fit ( ``nvidia_linux-4.6.4-93-367.be0`` or even ``nvidia_linux-4.6.4-93-current.be0``) and set the full version from inside.

> ends_with() lacks return value for failure

what failure? Its true or false. Did I overlook something?

> go into beeversion_parse.c and not be hacked into main

beeversion_parse.c:parse_version takes a string argument and that is a good building block. I think, the source of the string should stay in the callers domain. 

> your implementation would still need changes from this commit to work without regressions.

How can there be a regression? There is no change in processing for files not called ".be0", which didn't exist before.

> also only the first appearance "^#@beepackage" should "win" .. 

I think it does ( https://github.molgen.mpg.de/mariux64/bee/blob/master/src/beeversion.c#L164 )

**Comment by mariux** [comment]
>> but beeversion should check if package basename (full and normal) matches be0 filename

> Not wanted by us. Junk in, junk out is fine. But we might want to do that on purpose some day and don't want a "you cant do that" reply. Also, things get complicated with packages like ``nvidia_linux-4.6.4-93-367.18`` which beeversion parses to ``PKGFULLNAME=nvidia_linux-4.6.4-93`` When version/revision is stripped (to ``nvidia_linux-4.6.4-93``) it is parsed to `` PKGFULLNAME=nvidia_linux``. With ``nvidia_linux-4.6.4-93-367``, which makes some sense, its parsed to ``PKGFULLNAME=nvidia_linux-4.6.4`` which is also not what we want.  Without the check, we can ignore all this trouble. We can name the file as we see fit ( ``nvidia_linux-4.6.4-93-367.be0`` or even ``nvidia_linux-4.6.4-93-current.be0``) and set the full version from inside.

the apporach of checking package basename will work fine with packages having a really long pkgextraname including own versions.. junk in junk out is a source of errors.. it's all about expectation and consistency.. 

>> ends_with() lacks return value for failure
>what failure? Its true or false. Did I overlook something?

https://github.molgen.mpg.de/mariux64/bee/blob/master/src/beeversion.c#L125  no return code at the end!

>> your implementation would still need changes from this commit to work without regressions.

> How can there be a regression? There is no change in processing for files not called ".be0", which didn't exist before.

it does not check existing bee files for version conflicts for .be0 files.

**Comment by buczek** [comment]
> https://github.molgen.mpg.de/mariux64/bee/blob/master/src/beeversion.c#L125 no return code at the end!

OMG. Good point :-)

> it does not check existing bee files for version conflicts for .be0 files.

<pre>
root:theinternet:~/# /home/buczek/git/bee-files/gvfs.be0
BEE v1.2.10 2009-2016
  by  Marius Tolzmann <marius@mariux.de>
      Matthias Ruester <ruester@molgen.mpg.de>
      Tobias Dreyer <dreyer@molgen.mpg.de>

[BEE]   BEE_SKIPLIST           /etc/bee/skiplist
[BEE]   BEE_REPOSITORY_PREFIX  /src/mariux/beeroot
[BEE]   BEE_METADIR            /usr/share/bee
[BEE]   BEE_TMP_TMPDIR         /scratch/local
[BEE]   BEE_TMP_BUILDROOT      /scratch/local/bee-root
[BEE] package  already exists: /src/mariux/beeroot/packages/gvfs-1.24.0-2.x86_64.bee.tar.bz2
[BEE] please increase revision number
[BEE] cp /home/buczek/git/bee-files/gvfs.be0 /root/gvfs-1.24.0-3.bee
</pre>

Or do you mean something else?

**Comment by mariux** [comment]
`[BEE] bee-file already exists: ... `  is missing... because it needs at least the `check_rebuild` adjustment from this patch line...

**Comment by buczek** [comment]
I include the modified check_rebuild call, thanks.

**Pull request merged**